### PR TITLE
Reword callable reference and class literal resolution algorithm, explain some corner cases (#5)

### PR DESCRIPTION
**Pull request by udalov**

### CODE REVIEW

File: proposals/bound-callable-references.md [hunk 1]
```diff
@@ -123,20 +131,61 @@ fun test() {
...
+
+```
+class C {
+    companion object
+}
+
+fun test() {
+    C::class            // class of C
+    C()::class          // class of C
+    (C)::foo            // class of C.Companion
```
Comment (cypressious) [inline proposals/bound-callable-references.md]: shouldn't this be `::class`?

Comment (udalov) [inline proposals/bound-callable-references.md]: Fixed, thanks

File: proposals/bound-callable-references.md [hunk 2]
```diff
@@ -82,13 +86,17 @@ The LHS may now be interpreted as an expression, or a type, or both.
...
-The algorithm is the following:
+The semantics are different when the LHS is interpreted as an expression or a type, so we introduce the following algorithm to choose the resulting interpretation when both are applicable:
 
-1. Try interpreting the LHS as an **expression** with the usual resolution algorithm for qualified expressions.
-   If the result represents a companion object of some class, continue to p.2.
-   Otherwise continue resolution of the member in the scope of the expression's type.
-2. Resolve the unbound reference with the existing algorithm.
+1. Type-check the LHS as an **expression**.
+   - If the result represents a _companion object of some class_ specified with the short syntax (via the short/qualified name of the containing class), discard the result and continue to p.2.
+   - If the result represents an _object_ (either non-companion, or companion specified with the full syntax: `org.foo.Bar.Companion` or just `Bar.Companion`), remember the result and continue to p.2.
```
Comment (erokhins) [inline proposals/bound-callable-references.md]: or `(Bar)`

File: proposals/bound-callable-references.md [hunk 3]
```diff
@@ -82,13 +86,17 @@ The LHS may now be interpreted as an expression, or a type, or both.
...
-The semantics are different when the LHS is interpreted as an expression or a type, so we establish a priority of one over another when both interpretations are applicable.
-The algorithm is the following:
+The semantics are different when the LHS is interpreted as an expression or a type, so we introduce the following algorithm to choose the resulting interpretation when both are applicable:
 
-1. Try interpreting the LHS as an **expression** with the usual resolution algorithm for qualified expressions.
-   If the result represents a companion object of some class, continue to p.2.
-   Otherwise continue resolution of the member in the scope of the expression's type.
-2. Resolve the unbound reference with the existing algorithm.
+1. Type-check the LHS as an **expression**.
+   - If the result represents a _companion object of some class_ specified with the short syntax (via the short/qualified name of the containing class), discard the result and continue to p.2.
```
Comment (erokhins) [inline proposals/bound-callable-references.md]: represents a companion object **or object**. And what about enum entries?

File: proposals/bound-callable-references.md [hunk 4]
```diff
@@ -123,20 +131,61 @@ fun test() {
...
+Resolution of the LHS of a class literal expression is performed with the same algorithm.
+
+```
+class C {
+    companion object
+}
+
+fun test() {
+    C::class            // class of C
+    C()::class          // class of C
```
Comment (erokhins) [inline proposals/bound-callable-references.md]: `С()` class literal for instance?

File: proposals/bound-callable-references.md [hunk 5]
```diff
@@ -123,20 +131,61 @@ fun test() {
...
+    (C)::class          // class of C.Companion
+    C.Companion::class  // class of C.Companion
+}
+```
+
+Once the LHS is type-checked and its type is determined to be `T`, the type of the whole class literal expression is `kotlin.reflect.KClass<T'>` where `T'` is a type obtained by _substituting `T`'s arguments with star projections (`*`)_. Example:
```
Comment (erokhins) [inline proposals/bound-callable-references.md]: What about `Array<String>::class`?

File: proposals/bound-callable-references.md (hunk 2 above)
Comment (udalov) [inline proposals/bound-callable-references.md]: No, `(Bar)` is covered by the first clause because it's a companion object of some class specified with the _short_ syntax

File: proposals/bound-callable-references.md (hunk 3 above)
Comment (udalov) [inline proposals/bound-callable-references.md]: Why do you think non-companion objects are relevant here? They are handled in the second clause

File: proposals/bound-callable-references.md (hunk 5 above)
Comment (udalov) [inline proposals/bound-callable-references.md]: Yes, `arrayOfStrings::class` should have type `KClass<Array<String>>`. Thanks, I'll add this here

File: proposals/bound-callable-references.md (hunk 2 above)
Comment (erokhins) [inline proposals/bound-callable-references.md]: What about `object Foo` `(Foo)`?

Comment (udalov) [inline proposals/bound-callable-references.md]: I'm not sure what do you propose, could you elaborate? The two examples here explain what does "companion specified with the full syntax" mean, they do not relate to non-companion objects. I don't think usages of a non-companion object needs any explanation. Or are you talking about the `(Foo)::class` case?

File: proposals/bound-callable-references.md [hunk 6]
```diff
@@ -123,20 +131,61 @@ fun test() {
...
+```
+fun test() {
+    "a"::class               // KClass<String>
+    listOf("a")::class       // KClass<List<*>> (not "KClass<List<String>>"!)
```
Comment (abreslav) [inline proposals/bound-callable-references.md]: I'd add another example here: `listOf(listOf("a"))::class // KClass<List<*>>`

Otherwise, the reader might think that we only erase the innermost argument (I did think that for a second)... But maybe it's only me

**Comment by abreslav** [comment]
No other comments from me

**Comment by udalov** [comment]
@abreslav @erokhins I've updated the class literal type checking part with a more precise definition of what happens with expressions of type `kotlin.Array<...>`. Please review, I'm not sure I've formulated it well enough

File: proposals/bound-callable-references.md [hunk 7]
```diff
@@ -162,12 +162,24 @@ fun test() {
 }
 ```
 
-Once the LHS is type-checked and its type is determined to be `T`, the type of the whole class literal expression is `kotlin.reflect.KClass<T'>` where `T'` is a type obtained by _substituting `T`'s arguments with star projections (`*`)_. Example:
+Once the LHS is type-checked and its type is determined to be `T`, the type of the whole class literal expression is `KClass<erase(T)>`, where `erase(T)` represents the most accurate possible type-safe representation of the type `T` and is obtained by substituting `T`'s arguments with star projections (`*`), except when the type is an array.
```
Comment (abreslav) [inline proposals/bound-callable-references.md]: Type-arguments of arrays are not fully known at runtime either: their nullability is unclear. Does it matter in this context?

Comment (udalov) [inline proposals/bound-callable-references.md]: It does matter, and we may revisit it in the future, but type-checking `arrayOfStrings::class` to `KClass<Array<*>>` would be rather unhelpful from Java user's point of view. We've briefly discussed this with @erokhins and decided (for now) to type-check `arrayOfStrings::class` to `KClass<Array<String>>`. We should discuss this in detail later anyway.

Comment (bashor) [inline proposals/bound-callable-references.md]: Note in JS we don't know anything about type-arguments of arrays at runtime.

**Comment by udalov** [comment]
Added a TODO about JS, we'll have to discuss later. I'm merging this set of changes now, thanks to everyone for review!
